sdks/python/apache_beam/runners/portability/fn_api_runner.py

```diff
@@ -1286,30 +1375,36 @@ class WorkerHandlerManager(object):
   def __init__(self, environments, job_provision_info=None):
     self._environments = environments
     self._job_provision_info = job_provision_info
-    self._cached_handlers = {}
+    self._cached_handlers = collections.defaultdict(list)
     self._state = FnApiRunner.StateServicer() # rename?
+    # TODO(hannahjiang): create a grpc_server only for grpc env.
```

Comment:
We have to resolve this TODO before we merge, as by far the most common environment (e.g. for almost all our tests) is embedded.

Reply:
I am curious when we have multiple environments for a pipeline?

Reply:
Multi-language. (E.g. python might run embedded, go and java via docker.)